python/lsst/daf/butler/core/serverModels.py

```diff
@@ -82,33 +93,62 @@ def from_expression(cls, expression: Any) -> ExpressionQueryParameter:
             return cls()
 
         expressions = iterable(expression)
-        params: Dict[str, List[str]] = {}
+        params: Dict[str, List[str]] = {"glob": [], "regex": []}
         for expression in expressions:
             if expression is ...:
                 # This matches everything
                 return cls()
 
             if isinstance(expression, re.Pattern):
-                if (k := "regex") not in params:
-                    params[k] = []
                 params["regex"].append(expression.pattern)
-            else:
-                if (k := "glob") not in params:
-                    params[k] = []
+            elif isinstance(expression, str):
                 params["glob"].append(expression)
+            elif hasattr(expression, "name"):
+                params["glob"].append(expression.name)
```

Comment:
Level of duplicative logic is starting to worry me here, but I think there's a commit I haven't gotten to yet starts to address that...

Reply:
Oh, you are reviewing commit by commit. That is not a good idea on this pull request where much of the code is work in progress building up to something that works as I experiment and keep cleaning up as I learn more. I had not expected you to be reviewing like this.

Reply:
This doesn't seem like duplication though. It supports someone passing in a DatasetType but me wanting to send the name of that dataset type over the wire (and not the full DatasetType serialized form)

Reply:
> Oh, you are reviewing commit by commit. That is not a good idea on this pull request

Yeah, I realized that you didn't expect this last night while reading your replies, a couple of threads before you realized that's what I'd done.  In any case, aside from these false positives (I'm going through and resolving all of the clear-cut ones I see; feel free to resolve any others yourself), I think it may have actually helped me wrap my head around it a bit as it gave me a chronological view rather than an alphabetical view of a big suite of changes.

Reply:
And yes, I just misunderstood what this was doing r.e. duplication - it just looked suspiciously like the code that standardizes dataset type expressions as query inputs, but of course it's doing something quite different.

Reply:
although given your other ticket it sounds like I should be adding a `datasetTypes` field to the model and storing the full form for transmission to the server so it can work out exactly which dataset types it is going to be using and not just the name.

Reply:
Good point.  FWIW, if the user just passed the name you wouldn't need to pass more than that to the server (because the server could do the expansion), but if they did pass more, it would become important to respect that.  And it might make sense to always do the expansion on the client anyway.